In [1]:
import pandas as pd
import numpy as np 

In [2]:
!pip install geopy

     |████████████████████████████████| 112kB 8.6MB/s eta 0:00:01


In [3]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


<a id='item1'></a>

In [4]:
!pip install xlrd

     |████████████████████████████████| 112kB 23.8MB/s eta 0:00:01


In [8]:
df = pd.read_excel('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

XLRDError: Unsupported format, or corrupt file: Expected BOF record; found b'<!DOCTYP'

In [5]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [5]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

Let's take a quick look at the data.

In [6]:
newyork_data

{'type': 'FeatureCollection',
 'totalFeatures': 306,
 'features': [{'type': 'Feature',
   'id': 'nyu_2451_34572.1',
   'geometry': {'type': 'Point',
    'coordinates': [-73.84720052054902, 40.89470517661]},
   'geometry_name': 'geom',
   'properties': {'name': 'Wakefield',
    'stacked': 1,
    'annoline1': 'Wakefield',
    'annoline2': None,
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.84720052054902,
     40.89470517661,
     -73.84720052054902,
     40.89470517661]}},
  {'type': 'Feature',
   'id': 'nyu_2451_34572.2',
   'geometry': {'type': 'Point',
    'coordinates': [-73.82993910812398, 40.87429419303012]},
   'geometry_name': 'geom',
   'properties': {'name': 'Co-op City',
    'stacked': 2,
    'annoline1': 'Co-op',
    'annoline2': 'City',
    'annoline3': None,
    'annoangle': 0.0,
    'borough': 'Bronx',
    'bbox': [-73.82993910812398,
     40.87429419303012,
     -73.82993910812398,
     40.87429419303012]}},
  {'type': 'Feature',
 

In [7]:
neighborhoods_data = newyork_data['features']

In [8]:
neighborhoods_data[0]

{'type': 'Feature',
 'id': 'nyu_2451_34572.1',
 'geometry': {'type': 'Point',
  'coordinates': [-73.84720052054902, 40.89470517661]},
 'geometry_name': 'geom',
 'properties': {'name': 'Wakefield',
  'stacked': 1,
  'annoline1': 'Wakefield',
  'annoline2': None,
  'annoline3': None,
  'annoangle': 0.0,
  'borough': 'Bronx',
  'bbox': [-73.84720052054902,
   40.89470517661,
   -73.84720052054902,
   40.89470517661]}}

In [9]:
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
neighborhoods = pd.DataFrame(columns=column_names)

In [10]:
neighborhoods

Empty DataFrame
Columns: [Borough, Neighborhood, Latitude, Longitude]
Index: []

Then let's loop through the data and fill the dataframe one row at a time.

In [11]:
for data in neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough'] 
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    neighborhoods = neighborhoods.append({'Borough': borough,
                                          'Neighborhood': neighborhood_name,
                                          'Latitude': neighborhood_lat,
                                          'Longitude': neighborhood_lon}, ignore_index=True)

Quickly examine the resulting dataframe.

In [12]:
neighborhoods.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

And make sure that the dataset has all 5 boroughs and 306 neighborhoods.

In [13]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 5 boroughs and 306 neighborhoods.


#### Use geopy library to get the latitude and longitude values of New York City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [14]:
address = 'New York City, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.


#### Create a map of New York with neighborhoods superimposed on top.

In [15]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

In [16]:
manhattan_data = neighborhoods[neighborhoods['Borough'] == 'Manhattan'].reset_index(drop=True)
manhattan_data.head()

Borough        Neighborhood   Latitude  Longitude
0  Manhattan         Marble Hill  40.876551 -73.910660
1  Manhattan           Chinatown  40.715618 -73.994279
2  Manhattan  Washington Heights  40.851903 -73.936900
3  Manhattan              Inwood  40.867684 -73.921210
4  Manhattan    Hamilton Heights  40.823604 -73.949688

Let's get the geographical coordinates of Manhattan.

In [17]:
address = 'Manhattan, NY'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 40.7896239, -73.9598939.


As we did with all of New York City, let's visualizat Manhattan the neighborhoods in it.

In [18]:
# create map of Manhattan using latitude and longitude values
map_manhattan = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(manhattan_data['Latitude'], manhattan_data['Longitude'], manhattan_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_manhattan)  
    
map_manhattan

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [19]:
CLIENT_ID = 'OCEPD1EW4CBKDMFQEHKKZCZSZSKJSRJZFUUXUB321HXFSVPN' # your Foursquare ID
CLIENT_SECRET = '5CC4MQ22RTNZRVD4BSWFA5OC4MCULKTGDD4QXF5KOZH41CIE' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: OCEPD1EW4CBKDMFQEHKKZCZSZSKJSRJZFUUXUB321HXFSVPN
CLIENT_SECRET:5CC4MQ22RTNZRVD4BSWFA5OC4MCULKTGDD4QXF5KOZH41CIE


#### Let's explore the first neighborhood in our dataframe.

Get the neighborhood's name.

In [20]:
manhattan_data.loc[0, 'Neighborhood']

'Marble Hill'

Get the neighborhood's latitude and longitude values.

In [21]:
neighborhood_latitude = manhattan_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = manhattan_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = manhattan_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Marble Hill are 40.87655077879964, -73.91065965862981.


#### Now, let's get the top 100 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [22]:
# type your answer here

LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=OCEPD1EW4CBKDMFQEHKKZCZSZSKJSRJZFUUXUB321HXFSVPN&client_secret=5CC4MQ22RTNZRVD4BSWFA5OC4MCULKTGDD4QXF5KOZH41CIE&v=20180605&ll=40.87655077879964,-73.91065965862981&radius=500&limit=100'

Double-click __here__ for the solution.
<!-- The correct answer is:
LIMIT = 100 # limit of number of venues returned by Foursquare API
-->

<!--
radius = 500 # define radius
-->

<!--
\\ # create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
url # display URL
--> 

Send the GET request and examine the resutls

In [23]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e5b94ede826ac001e44ccc5'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Marble Hill',
  'headerFullLocation': 'Marble Hill, New York',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 24,
  'suggestedBounds': {'ne': {'lat': 40.88105078329964,
    'lng': -73.90471933917806},
   'sw': {'lat': 40.87205077429964, 'lng': -73.91659997808156}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4b4429abf964a52037f225e3',
       'name': "Arturo's",
       'location': {'address': '5198 Broadway',
        'crossStreet': 'at 225th St.',
        'lat': 40.87441177110231,
        'lng': -73.91027100981574,
        'labeledLatLngs': [{'label'

From the Foursquare lab in the previous module, we know that all the information is in the *items* key. Before we proceed, let's borrow the **get_category_type** function from the Foursquare lab.

In [24]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [25]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name   categories        lat        lng
0       Arturo's  Pizza Place  40.874412 -73.910271
1    Bikram Yoga  Yoga Studio  40.876844 -73.906204
2  Tibbett Diner        Diner  40.880404 -73.908937
3        Dunkin'   Donut Shop  40.877136 -73.906666
4      Starbucks  Coffee Shop  40.877531 -73.905582

And how many venues were returned by Foursquare?

In [26]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

24 venues were returned by Foursquare.


<a id='item2'></a>

## 2. Explore Neighborhoods in Manhattan

#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan

In [27]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *manhattan_venues*.

In [28]:
# type your answer here

manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )



Marble Hill
Chinatown
Washington Heights
Inwood
Hamilton Heights
Manhattanville
Central Harlem
East Harlem
Upper East Side
Yorkville
Lenox Hill
Roosevelt Island
Upper West Side
Lincoln Square
Clinton
Midtown
Murray Hill
Chelsea
Greenwich Village
East Village
Lower East Side
Tribeca
Little Italy
Soho
West Village
Manhattan Valley
Morningside Heights
Gramercy
Battery Park City
Financial District
Carnegie Hill
Noho
Civic Center
Midtown South
Sutton Place
Turtle Bay
Tudor City
Stuyvesant Town
Flatiron
Hudson Yards


Double-click __here__ for the solution.
<!-- The correct answer is:
manhattan_venues = getNearbyVenues(names=manhattan_data['Neighborhood'],
                                   latitudes=manhattan_data['Latitude'],
                                   longitudes=manhattan_data['Longitude']
                                  )
--> 

#### Let's check the size of the resulting dataframe

In [29]:
print(manhattan_venues.shape)
manhattan_venues.head()

(3320, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude          Venue  \
0  Marble Hill              40.876551               -73.91066       Arturo's   
1  Marble Hill              40.876551               -73.91066    Bikram Yoga   
2  Marble Hill              40.876551               -73.91066  Tibbett Diner   
3  Marble Hill              40.876551               -73.91066        Dunkin'   
4  Marble Hill              40.876551               -73.91066      Starbucks   

   Venue Latitude  Venue Longitude Venue Category  
0       40.874412       -73.910271    Pizza Place  
1       40.876844       -73.906204    Yoga Studio  
2       40.880404       -73.908937          Diner  
3       40.877136       -73.906666     Donut Shop  
4       40.877531       -73.905582    Coffee Shop

Let's check how many venues were returned for each neighborhood

In [30]:
manhattan_venues.groupby('Neighborhood').count()

Neighborhood Latitude  Neighborhood Longitude  Venue  \
Neighborhood                                                                
Battery Park City                       98                      98     98   
Carnegie Hill                           99                      99     99   
Central Harlem                          46                      46     46   
Chelsea                                100                     100    100   
Chinatown                              100                     100    100   
Civic Center                           100                     100    100   
Clinton                                100                     100    100   
East Harlem                             42                      42     42   
East Village                           100                     100    100   
Financial District                     100                     100    100   
Flatiron                               100                     100    100   
Gramercy                               100                     100    100   
Greenwich Village                      100                     100    100   
Hamilton Heights                        59                      59     59   
Hudson Yards                            81                      81     81   
Inwood                                  58                      58     58   
Lenox Hill                             100                     100    100   
Lincoln Square                         100                     100    100   
Little Italy                           100                     100    100   
Lower East Side                         60                      60     60   
Manhattan Valley                        53                      53     53   
Manhattanville                          42                      42     42   
Marble Hill                             24                      24     24   
Midtown                                100                     100    100   
Midtown South                          100                     100    100   
Morningside Heights                     43                      43     43   
Murray Hill                            100                     100    100   
Noho                                   100                     100    100   
Roosevelt Island                        26                      26     26   
Soho                                   100                     100    100   
Stuyvesant Town                         19                      19     19   
Sutton Place                           100                     100    100   
Tribeca                                100                     100    100   
Tudor City                              82                      82     82   
Turtle Bay                             100                     100    100   
Upper East Side                        100                     100    100   
Upper West Side                        100                     100    100   
Washington Heights                      88                      88     88   
West Village                           100                     100    100   
Yorkville                              100                     100    100   

                     Venue Latitude  Venue Longitude  Venue Category  
Neighborhood                                                          
Battery Park City                98               98              98  
Carnegie Hill                    99               99              99  
Central Harlem                   46               46              46  
Chelsea                         100              100             100  
Chinatown                       100              100             100  
Civic Center                    100              100             100  
Clinton                         100              100             100  
East Harlem                      42               42              42  
East Village                    100              100             100  
Finan

#### Let's find out how many unique categories can be curated from all the returned venues

In [31]:
print('There are {} uniques categories.'.format(len(manhattan_venues['Venue Category'].unique())))

There are 337 uniques categories.


<a id='item3'></a>

## 3. Analyze Each Neighborhood

In [32]:
# one hot encoding
manhattan_onehot = pd.get_dummies(manhattan_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
manhattan_onehot['Neighborhood'] = manhattan_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [manhattan_onehot.columns[-1]] + list(manhattan_onehot.columns[:-1])
manhattan_onehot = manhattan_onehot[fixed_columns]

manhattan_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0  Marble Hill                  0               0                  0   
1  Marble Hill                  0               0                  0   
2  Marble Hill                  0               0                  0   
3  Marble Hill                  0               0                  0   
4  Marble Hill                  0               0                  0   

   African Restaurant  American Restaurant  Antique Shop  Arcade  \
0                   0                    0             0       0   
1                   0                    0             0       0   
2                   0                    0             0       0   
3                   0                    0             0       0   
4                   0                    0             0       0   

   Arepa Restaurant  Argentinian Restaurant  Art Gallery  Art Museum  \
0                 0                       0            0           0   
1                 0                       0            0           0   
2                 0                       0            0           0   
3                 0                       0            0           0   
4                 0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auditorium  \
0                    0                 0                   0           0   
1                    0                 0                   0           0   
2                    0                 0                   0           0   
3                    0                 0                   0           0   
4                    0                 0                   0           0   

   Australian Restaurant  Austrian Restaurant  BBQ Joint  Baby Store  \
0                      0                    0          0           0   
1                      0                    0          0           0   
2                      0                    0          0           0   
3                      0                    0          0           0   
4                      0                    0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Basketball Court  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                0         0            0           0              0   
1                0         0            0           0              0   
2                0         0            0           0              0   
3                0         0            0           0              0   
4                0         0            0           0              0   

   Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  Board Shop  \
0                         0          0           0       0           0   
1                         0          0           0       0           0   
2                         0          0           0       0           0   
3                         0          0           0       0           0   
4                         0          0           0       0           0   

   Boat or Ferry  Bookstore  Boutique  Boxing Gym  Brazilian Restaurant  \
0              0          0         0           0                     0   
1              0          0         0           0                     0   
2              0          0         0           0                     0   
3              0          0         0           0                     0   
4              0          0         0           0                     0   

   Breakfast Spot  Bridal Shop  Bridge  Bubble Tea Shop  Building  \
0               0 

And let's examine the new dataframe size.

In [33]:
manhattan_onehot.shape

(3320, 338)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [34]:
manhattan_grouped = manhattan_onehot.groupby('Neighborhood').mean().reset_index()
manhattan_grouped

Neighborhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Battery Park City           0.000000            0.00               0.00   
1         Carnegie Hill           0.000000            0.00               0.00   
2        Central Harlem           0.000000            0.00               0.00   
3               Chelsea           0.000000            0.00               0.00   
4             Chinatown           0.000000            0.00               0.00   
5          Civic Center           0.000000            0.00               0.00   
6               Clinton           0.000000            0.00               0.00   
7           East Harlem           0.000000            0.00               0.00   
8          East Village           0.000000            0.00               0.00   
9    Financial District           0.010000            0.00               0.00   
10             Flatiron           0.000000            0.00               0.00   
11             Gramercy           0.000000            0.00               0.00   
12    Greenwich Village           0.000000            0.00               0.00   
13     Hamilton Heights           0.000000            0.00               0.00   
14         Hudson Yards           0.000000            0.00               0.00   
15               Inwood           0.000000            0.00               0.00   
16           Lenox Hill           0.000000            0.00               0.01   
17       Lincoln Square           0.000000            0.00               0.00   
18         Little Italy           0.000000            0.00               0.00   
19      Lower East Side           0.000000            0.00               0.00   
20     Manhattan Valley           0.000000            0.00               0.00   
21       Manhattanville           0.000000            0.00               0.00   
22          Marble Hill           0.000000            0.00               0.00   
23              Midtown           0.000000            0.00               0.00   
24        Midtown South           0.000000            0.00               0.00   
25  Morningside Heights           0.000000            0.00               0.00   
26          Murray Hill           0.000000            0.00               0.00   
27                 Noho           0.000000            0.01               0.00   
28     Roosevelt Island           0.000000            0.00               0.00   
29                 Soho           0.010000            0.00               0.00   
30      Stuyvesant Town           0.000000            0.00               0.00   
31         Sutton Place           0.000000            0.01               0.00   
32              Tribeca           0.000000            0.00               0.00   
33           Tudor City           0.000000            0.00               0.00   
34           Turtle Bay           0.000000            0.00               0.00   
35      Upper East Side           0.000000            0.00               0.00   
36      Upper West Side           0.010000            0.00               0.00   
37   Washington Heights           0.011364            0.00               0.00   
38         West Village           0.010000            0.00               0.00   
39            Yorkville           0.000000            0.00               0.00   

    African Restaurant  American Restaurant  Antique Shop  Arcade  \
0             0.000000             0.010204          0.00    0.00   
1             0.000000             0.010101          0.00    0.00   
2             0.043478             0.043478          0.00    0.00   
3             0.000000             0.030000          0.00    0.00   
4             0.000000             0.040000          0.00    0.00   
5             0.000000             0.030000          0.01    0.00   
6             0.000000             0.040000          0.00    0.00   
7             0.000000             0.000000          0.00    0.00   
8             0.000000             0.020000          0.00    0.00   

#### Let's confirm the new size

In [35]:
manhattan_grouped.shape

(40, 338)

#### Let's print each neighborhood along with the top 5 most common venues

In [36]:
num_top_venues = 5

for hood in manhattan_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = manhattan_grouped[manhattan_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Battery Park City----
           venue  freq
0    Coffee Shop  0.07
1           Park  0.07
2          Hotel  0.05
3            Gym  0.04
4  Shopping Mall  0.03


----Carnegie Hill----
               venue  freq
0        Coffee Shop  0.07
1        Pizza Place  0.05
2        Yoga Studio  0.03
3  French Restaurant  0.03
4               Café  0.03


----Central Harlem----
                 venue  freq
0   Chinese Restaurant  0.07
1                  Bar  0.04
2   African Restaurant  0.04
3  American Restaurant  0.04
4    French Restaurant  0.04


----Chelsea----
                venue  freq
0         Coffee Shop  0.06
1              Bakery  0.05
2  Italian Restaurant  0.05
3      Ice Cream Shop  0.03
4           Nightclub  0.03


----Chinatown----
                   venue  freq
0     Chinese Restaurant  0.09
1           Cocktail Bar  0.05
2  Vietnamese Restaurant  0.04
3    American Restaurant  0.04
4      Hotpot Restaurant  0.03


----Civic Center----
                  venue  freq
0     

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [37]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [38]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = manhattan_grouped['Neighborhood']

for ind in np.arange(manhattan_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(manhattan_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Battery Park City                  Park           Coffee Shop   
1      Carnegie Hill           Coffee Shop           Pizza Place   
2     Central Harlem    Chinese Restaurant        Cosmetics Shop   
3            Chelsea           Coffee Shop                Bakery   
4          Chinatown    Chinese Restaurant          Cocktail Bar   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Hotel                   Gym         Women's Store   
1            Yoga Studio                Bakery             Bookstore   
2      French Restaurant    African Restaurant   American Restaurant   
3     Italian Restaurant             Nightclub               Theater   
4  Vietnamese Restaurant   American Restaurant          Optical Shop   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Shopping Mall         Memorial Site         Boat or Ferry   
1                  Café        Cosmetics Shop     French Restaurant   
2                   Bar    Seafood Restaurant             Bookstore   
3           Art Gallery                 Hotel        Ice Cream Shop   
4     Hotpot Restaurant                   Spa    Dim Sum Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0           Men's Store              BBQ Joint  
1         Grocery Store                    Gym  
2              Beer Bar   Gym / Fitness Center  
3   American Restaurant              Wine Shop  
4    Salon / Barbershop                 Bakery

<a id='item4'></a>

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [39]:
# set number of clusters
kclusters = 5

manhattan_grouped_clustering = manhattan_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(manhattan_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 3, 0, 0, 0, 0, 0, 4, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [40]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

manhattan_merged = manhattan_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
manhattan_merged = manhattan_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

manhattan_merged.head() # check the last columns!

Borough        Neighborhood   Latitude  Longitude  Cluster Labels  \
0  Manhattan         Marble Hill  40.876551 -73.910660               4   
1  Manhattan           Chinatown  40.715618 -73.994279               0   
2  Manhattan  Washington Heights  40.851903 -73.936900               4   
3  Manhattan              Inwood  40.867684 -73.921210               4   
4  Manhattan    Hamilton Heights  40.823604 -73.949688               4   

  1st Most Common Venue 2nd Most Common Venue  3rd Most Common Venue  \
0                   Gym        Sandwich Place            Coffee Shop   
1    Chinese Restaurant          Cocktail Bar  Vietnamese Restaurant   
2                  Café         Grocery Store                 Bakery   
3    Mexican Restaurant           Pizza Place                 Lounge   
4           Pizza Place         Deli / Bodega            Coffee Shop   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0           Yoga Studio         Deli / Bodega       Supplement Shop   
1   American Restaurant          Optical Shop     Hotpot Restaurant   
2     Mobile Phone Shop         Deli / Bodega                   Gym   
3                  Café            Restaurant         Deli / Bodega   
4                  Café    Mexican Restaurant                Bakery   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Steakhouse         Shopping Mall    Seafood Restaurant   
1                   Spa    Dim Sum Restaurant    Salon / Barbershop   
2                  Bank    Chinese Restaurant      Tapas Restaurant   
3    Spanish Restaurant                  Park                Bakery   
4                  Park        Sandwich Place          Cocktail Bar   

  10th Most Common Venue  
0            Pizza Place  
1                 Bakery  
2             Restaurant  
3     Frozen Yogurt Shop  
4                 School

Finally, let's visualize the resulting clusters

In [41]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(manhattan_merged['Latitude'], manhattan_merged['Longitude'], manhattan_merged['Neighborhood'], manhattan_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id='item5'></a>

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1

In [42]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 0, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue    2nd Most Common Venue  \
1           Chinatown    Chinese Restaurant             Cocktail Bar   
6      Central Harlem    Chinese Restaurant           Cosmetics Shop   
8     Upper East Side               Exhibit       Italian Restaurant   
13     Lincoln Square                 Plaza       Italian Restaurant   
14            Clinton               Theater       Italian Restaurant   
17            Chelsea           Coffee Shop                   Bakery   
18  Greenwich Village    Italian Restaurant           Clothing Store   
21            Tribeca   American Restaurant                     Park   
24       West Village    Italian Restaurant  New American Restaurant   
31               Noho    Italian Restaurant        French Restaurant   
32       Civic Center           Coffee Shop     Gym / Fitness Center   
34       Sutton Place  Gym / Fitness Center                      Gym   
35         Turtle Bay    Italian Restaurant               Steakhouse   
39       Hudson Yards   American Restaurant     Gym / Fitness Center   

     3rd Most Common Venue 4th Most Common Venue  5th Most Common Venue  \
1    Vietnamese Restaurant   American Restaurant           Optical Shop   
6        French Restaurant    African Restaurant    American Restaurant   
8              Art Gallery                Bakery            Coffee Shop   
13                    Café               Theater  Performing Arts Venue   
14    Gym / Fitness Center           Coffee Shop    American Restaurant   
17      Italian Restaurant             Nightclub                Theater   
18        Sushi Restaurant                  Café      Indian Restaurant   
21      Italian Restaurant                   Spa                   Café   
24          Cosmetics Shop   American Restaurant           Cocktail Bar   
31             Coffee Shop                 Hotel           Cocktail Bar   
32                   Hotel     French Restaurant            Yoga Studio   
34  Furniture / Home Store    Italian Restaurant      Indian Restaurant   
35             Coffee Shop      Sushi Restaurant               Wine Bar   
39                   Hotel    Italian Restaurant            Coffee Shop   

   6th Most Common Venue      7th Most Common Venue 8th Most Common Venue  \
1      Hotpot Restaurant                        Spa    Dim Sum Restaurant   
6                    Bar         Seafood Restaurant             Bookstore   
8   Gym / Fitness Center                  Juice Bar     French Restaurant   
13          Concert Hall                       Park  Gym / Fitness Center   
14        Sandwich Place                      Hotel          Cocktail Bar   
17           Art Gallery                      Hotel        Ice Cream Shop   
18     French Restaurant         Seafood Restaurant          Burger Joint   
21             Wine Shop                   Wine Bar      Greek Restaurant   
24              Wine Bar                       Park               Theater   
31   American Restaurant              Grocery Store           Art Gallery   
32    Italian Restaurant             Sandwich Place          Cocktail Bar   
34           Coffee Shop  Latin American Restaurant     French Restaurant   
35                  Café           Ramen Restaurant     French Restaurant   
39       Thai Restaurant                       Café                   Gym   

      9th Most Common Venue 10th Most Common Venue  
1        Salon / Barbershop                 Bakery  
6                  Beer Bar   Gym / Fitness Center  
8                     Hotel               Boutique  
13      American Restaurant    Indie Movie Theater  
14  New American Restaurant                    Gym  
17      American Restaurant              Wine Shop  
18           Cosmetics Shop           Dessert Shop  
21              Coffee Shop             Playground  
24                   Bakery         Ice Cream Shop  
31              Pizza Place              Gift Shop  
32                     Park                    Spa  
34        

#### Cluster 2

In [43]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 1, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
9            Yorkville    Italian Restaurant                   Bar   
10          Lenox Hill    Italian Restaurant           Coffee Shop   
12     Upper West Side    Italian Restaurant              Wine Bar   
19        East Village                   Bar        Ice Cream Shop   
25    Manhattan Valley                   Bar           Pizza Place   
27            Gramercy                   Bar    Italian Restaurant   
29  Financial District           Coffee Shop   American Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
9                    Gym           Coffee Shop         Deli / Bodega   
10      Sushi Restaurant           Pizza Place          Cocktail Bar   
12                   Bar                  Café           Coffee Shop   
19    Chinese Restaurant           Pizza Place    Mexican Restaurant   
25     Indian Restaurant           Coffee Shop    Mexican Restaurant   
27            Bagel Shop       Thai Restaurant           Pizza Place   
29           Pizza Place                 Hotel                   Gym   

   6th Most Common Venue 7th Most Common Venue          8th Most Common Venue  \
9            Pizza Place             Wine Shop             Mexican Restaurant   
10          Burger Joint                   Gym                           Café   
12     Indian Restaurant                Bakery  Vegetarian / Vegan Restaurant   
19              Wine Bar          Cocktail Bar            Japanese Restaurant   
25            Playground           Yoga Studio                 Clothing Store   
27           Coffee Shop          Cocktail Bar             Mexican Restaurant   
29                   Bar  Gym / Fitness Center                     Food Truck   

   9th Most Common Venue         10th Most Common Venue  
9    Japanese Restaurant               Sushi Restaurant  
10  Gym / Fitness Center             Salon / Barbershop  
12  Gym / Fitness Center       Mediterranean Restaurant  
19    Italian Restaurant  Vegetarian / Vegan Restaurant  
25    Chinese Restaurant                           Park  
27           Comedy Club         Thrift / Vintage Store  
29            Steakhouse             Italian Restaurant

#### Cluster 3

In [44]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 2, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
37  Stuyvesant Town         Boat or Ferry                  Park   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
37                   Bar        Baseball Field           Pet Service   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
37           Gas Station        Farmers Market     German Restaurant   

   9th Most Common Venue 10th Most Common Venue  
37  Gym / Fitness Center        Harbor / Marina

#### Cluster 4

In [45]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 3, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
15            Midtown                 Hotel        Clothing Store   
16        Murray Hill        Sandwich Place                 Hotel   
22       Little Italy                  Café                Bakery   
23               Soho        Clothing Store              Boutique   
28  Battery Park City                  Park           Coffee Shop   
30      Carnegie Hill           Coffee Shop           Pizza Place   
33      Midtown South     Korean Restaurant                 Hotel   
38           Flatiron           Yoga Studio  Gym / Fitness Center   

   3rd Most Common Venue 4th Most Common Venue     5th Most Common Venue  \
15           Coffee Shop   Sporting Goods Shop                    Bakery   
16   American Restaurant           Coffee Shop       Japanese Restaurant   
22       Bubble Tea Shop    Salon / Barbershop  Mediterranean Restaurant   
23            Shoe Store         Women's Store               Art Gallery   
28                 Hotel                   Gym             Women's Store   
30           Yoga Studio                Bakery                 Bookstore   
33             Hotel Bar   Japanese Restaurant              Dessert Shop   
38                  Café   American Restaurant       Japanese Restaurant   

       6th Most Common Venue     7th Most Common Venue  \
15              Cocktail Bar                Steakhouse   
16  Mediterranean Restaurant        Italian Restaurant   
22            Sandwich Place        Italian Restaurant   
23       Sporting Goods Shop    Furniture / Home Store   
28             Shopping Mall             Memorial Site   
30                      Café            Cosmetics Shop   
33      Gym / Fitness Center               Coffee Shop   
38            Clothing Store  Mediterranean Restaurant   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
15                Bookstore                  Café                    Gym  
16                      Gym  Gym / Fitness Center      Jewish Restaurant  
22             Cocktail Bar        Ice Cream Shop      French Restaurant  
23                   Bakery    Italian Restaurant            Men's Store  
28            Boat or Ferry           Men's Store              BBQ Joint  
30        French Restaurant         Grocery Store                    Gym  
33      American Restaurant        Cosmetics Shop         Scenic Lookout  
38  New American Restaurant                   Spa         Cosmetics Shop

#### Cluster 5

In [46]:
manhattan_merged.loc[manhattan_merged['Cluster Labels'] == 4, manhattan_merged.columns[[1] + list(range(5, manhattan_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0           Marble Hill                   Gym        Sandwich Place   
2    Washington Heights                  Café         Grocery Store   
3                Inwood    Mexican Restaurant           Pizza Place   
4      Hamilton Heights           Pizza Place         Deli / Bodega   
5        Manhattanville         Deli / Bodega           Coffee Shop   
7           East Harlem    Mexican Restaurant       Thai Restaurant   
11     Roosevelt Island        Sandwich Place       Bubble Tea Shop   
20      Lower East Side                  Café           Pizza Place   
26  Morningside Heights                  Park   American Restaurant   
36           Tudor City    Mexican Restaurant                  Café   

        3rd Most Common Venue 4th Most Common Venue    5th Most Common Venue  \
0                 Coffee Shop           Yoga Studio            Deli / Bodega   
2                      Bakery     Mobile Phone Shop            Deli / Bodega   
3                      Lounge                  Café               Restaurant   
4                 Coffee Shop                  Café       Mexican Restaurant   
5          Italian Restaurant    Mexican Restaurant       Seafood Restaurant   
7   Latin American Restaurant                Bakery            Deli / Bodega   
11           Greek Restaurant           Dry Cleaner                     Park   
20                Coffee Shop           Art Gallery       Chinese Restaurant   
26                Coffee Shop             Bookstore  New American Restaurant   
36                       Park         Deli / Bodega              Pizza Place   

   6th Most Common Venue      7th Most Common Venue 8th Most Common Venue  \
0        Supplement Shop                 Steakhouse         Shopping Mall   
2                    Gym                       Bank    Chinese Restaurant   
3          Deli / Bodega         Spanish Restaurant                  Park   
4                 Bakery                       Park        Sandwich Place   
5                   Park  Japanese Curry Restaurant      Cuban Restaurant   
7                   Café                        Spa         Grocery Store   
11           Coffee Shop                        Gym          Liquor Store   
20              Pharmacy           Ramen Restaurant                Bakery   
26          Burger Joint              Deli / Bodega        Sandwich Place   
36           Coffee Shop                      Diner      Greek Restaurant   

   9th Most Common Venue    10th Most Common Venue  
0     Seafood Restaurant               Pizza Place  
2       Tapas Restaurant                Restaurant  
3                 Bakery        Frozen Yogurt Shop  
4           Cocktail Bar                    School  
5            Supermarket                  Boutique  
7   Fast Food Restaurant               Coffee Shop  
11        Scenic Lookout                   Dog Run  
20   Japanese Restaurant              Cocktail Bar  
26        Ice Cream Shop  Mediterranean Restaurant  
36               Dog Run        Spanish Restaurant